File: api/src/main/java/org/delivery/api/domain/token/helper/JwtTokenHelper.java
package org.delivery.api.domain.token.helper;


import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jws;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import io.jsonwebtoken.security.SignatureException;
import java.sql.Date;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.HashMap;
import java.util.Map;
import org.delivery.common.error.ErrorCode;
import org.delivery.common.error.TokenErrorCode;
import org.delivery.common.exception.ApiException;
import org.delivery.api.domain.token.ifs.TokenHelperInterface;
import org.delivery.api.domain.token.model.TokenDto;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component
public class JwtTokenHelper implements TokenHelperInterface {

    @Value("${token.secret.key}") // Spring에서 application.yaml에 적힌 내용을 통해 자동으로 값을 넣어준다.
    private String secretKey;

    @Value("${token.access-token.plus-hour}")
    private Long accessTokenPlusHour;

    @Value("${token.refresh-token.plus-hour}")
    private Long refreshTokenPlusHour;

    @Override
    public TokenDto issueAccessToken(Map<String, Object> data) {
        var expiredLocalDateTime = LocalDateTime.now().plusHours(accessTokenPlusHour);
        var expiredAt = Date.from(
                expiredLocalDateTime.atZone(
                        ZoneId.systemDefault()
                ).toInstant()
        );

        var key = Keys.hmacShaKeyFor(secretKey.getBytes());

        var jwtToken = Jwts.builder()
                .signWith(key, SignatureAlgorithm.HS256)
                .setClaims(data)
                .setExpiration(expiredAt)
                .compact();

        return TokenDto.builder()
                .token(jwtToken)
                .expiredAt(expiredLocalDateTime)
                .build();
    }

    @Override
    public TokenDto issueRefreshToken(Map<String, Object> data) {
        var expiredLocalDateTime = LocalDateTime.now().plusHours(refreshTokenPlusHour);
        var expiredAt = Date.from(
                expiredLocalDateTime.atZone(
                        ZoneId.systemDefault()
                ).toInstant()
        );

        var key = Keys.hmacShaKeyFor(secretKey.getBytes());

        var jwtToken = Jwts.builder()
                .signWith(key, SignatureAlgorithm.HS256)
                .setClaims(data)
                .setExpiration(expiredAt)
                .compact();

        return TokenDto.builder()
                .token(jwtToken)
                .expiredAt(expiredLocalDateTime)
                .build();
    }

    @Override
    public Map<String, Object> validationTokenWithThrow(String token) {
        var key = Keys.hmacShaKeyFor(secretKey.getBytes());

        var parser = Jwts.parserBuilder()
                .setSigningKey(key)
                .build();

        try {

            Jws<Claims> result = parser.parseClaimsJws(token);
            return new HashMap<String, Object>(result.getBody());

        } catch (Exception e) {

            if (e instanceof SignatureException) { // 토큰이 유효하지 않을 때
                throw new ApiException(TokenErrorCode.INVALID_TOKEN, e);
            } else if (e instanceof ExpiredJwtException) { // 토큰이 만료되었을 때
                throw new ApiException(TokenErrorCode.EXPIRED_TOKEN, e);
            } else { // 그 외의 경우
                throw new ApiException(TokenErrorCode.TOKEN_EXCEPTION, e);
            }

        }
    }
}

File: account/src/main/kotlin/org/delivery/account/domain/token/helper/JwtTokenHelper.kt
package org.delivery.account.domain.token.helper

import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.security.Keys
import io.jsonwebtoken.security.SignatureException
import org.delivery.account.domain.token.ifs.TokenHelperInterface
import org.delivery.account.domain.token.model.TokenDto
import org.delivery.common.error.TokenErrorCode
import org.delivery.common.exception.ApiException
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.lang.Exception
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Date

@Component
class JwtTokenHelper: TokenHelperInterface{

    @Value("\${token.secret.key}") // Spring에서 application.yaml에 적힌 내용을 통해 자동으로 값을 넣어준다.
    private val secretKey: String? = null

    @Value("\${token.access-token.plus-hour}")
    private val accessTokenPlusHour: Long = 1

    @Value("\${token.refresh-token.plus-hour}")
    private val refreshTokenPlusHour: Long = 12

    override fun issueAccessToken(data: Map<String, Any>?): TokenDto? {

        val expiredLocalDateTime = LocalDateTime.now().plusHours(accessTokenPlusHour)

        val expiredAt = Date.from(expiredLocalDateTime.atZone(ZoneId.systemDefault()).toInstant())

        val key = Keys.hmacShaKeyFor(secretKey?.toByteArray())

        val jwtToken = Jwts.builder()
            .signWith(key, SignatureAlgorithm.HS256)
            .setClaims(data)
            .setExpiration(expiredAt)
            .compact()

        return TokenDto(token = jwtToken, expiredAt = expiredLocalDateTime)

    }

    override fun issueRefreshToken(data: Map<String, Any>?): TokenDto? {

        val expiredLocalDateTime = LocalDateTime.now().plusHours(refreshTokenPlusHour)

        val expiredAt = Date.from(expiredLocalDateTime.atZone(ZoneId.systemDefault()).toInstant())

        val key = Keys.hmacShaKeyFor(secretKey?.toByteArray())

        val jwtToken = Jwts.builder()
            .signWith(key, SignatureAlgorithm.HS256)
            .setClaims(data)
            .setExpiration(expiredAt)
            .compact()

        return TokenDto(token = jwtToken, expiredAt = expiredLocalDateTime)

    }

    override fun validationTokenWithThrow(token: String?): Map<String, Any>? {

        val key = Keys.hmacShaKeyFor(secretKey?.toByteArray())

        val parser = Jwts.parserBuilder()
            .setSigningKey(key)
            .build()

        return try {
            val result = token?.let { parser.parseClaimsJws(it) }
            HashMap(result?.body) // 가장 마지막 라인이 리턴
        } catch (e: Exception) {
            when(e) {
                is SignatureException -> {
                    throw ApiException(TokenErrorCode.INVALID_TOKEN)
                }
                is ExpiredJwtException -> {
                    throw ApiException(TokenErrorCode.EXPIRED_TOKEN)
                }
                else -> {
                    throw ApiException(TokenErrorCode.TOKEN_EXCEPTION)
                }
            }
        }

    }
}